Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [196]:
NAME = "Shadab Iqbal"
COLLABORATORS = ""

---

# First import the necessary libraries

In [197]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn import datasets, preprocessing, linear_model
from sklearn.model_selection import train_test_split

# Linear Regression

In [198]:
# Import the diabetes dataset from sklearn's datasets module, save it to the variable data
data = datasets.load_diabetes()

# Split the data into X and y. Print the shape of the two array.
X = data['data']
y = data['target']

print(X.shape)
print(y.shape)

(442, 10)
(442,)


## Question 1
How many samples are there in this datasets? How many features are there per sample?

YOUR ANSWER HERE:

There are 442 samples in this database and the number of features per sample is 10.

In [199]:
# Display the first 5 rows of the dataset in a pandas dataframe format

# YOUR CODE HERE
df = pd.DataFrame(X, columns=data["feature_names"])
df['target'] = y
df.head()

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  target  
0 -0.002592  0.019908 -0.017646   151.0  
1 -0.039493 -0.068330 -0.092204    75.0  
2 -0.002592  0.002864 -0.025930   141.0  
3  0.034309  0.022692 -0.009362   206.0  
4 -0.002592 -0.031991 -0.046641   135.0

In [200]:
# Print the maximum and the minimum of each each feature

# YOUR CODE HERE
print("Feature wise maxima")
print(df.max())
print("Feature wise minima")
print(df.min())

Feature wise maxima
age         0.110727
sex         0.050680
bmi         0.170555
bp          0.132044
s1          0.153914
s2          0.198788
s3          0.181179
s4          0.185234
s5          0.133599
s6          0.135612
target    346.000000
dtype: float64
Feature wise minima
age       -0.107226
sex       -0.044642
bmi       -0.090275
bp        -0.112400
s1        -0.126781
s2        -0.115613
s3        -0.102307
s4        -0.076395
s5        -0.126097
s6        -0.137767
target    25.000000
dtype: float64


## Question 2
Are the minimums and the maximums close to each other? Hence, do we need any normalization or standardization?

YOUR ANSWER HERE

Yes, the maxima and minima are close enough to each other. Hence we DO NOT need any normalization.

In [201]:
# Split the data into training and testing set, with 20% of the data in the testing set. Use random_state = 100
# Then print the shapes of X_train and X_test to see how many samples are on each set

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=100)

# YOUR CODE HERE
print(X_train.shape)
print(X_test.shape)

(353, 10)
(89, 10)


In [202]:
# Fit a linear regression model on X_train, y_train
# predict the model's output for X_train and save it to y_predict_train
# predict the model's output for X_test and save it to y_predict_test

# YOUR CODE HERE
model = linear_model.LinearRegression().fit(X_train, y_train)
y_predict_train = model.predict(X_train)
y_predict_test = model.predict(X_test)
# print(y_predict_test)

In [203]:
# Calculate a suitable performance metric for the training set for this regression problem
# Calculate a suitable performance metric for the test set for this regression problem
# Print the values

performance_train = np.mean(abs(y_train - y_predict_train))
performance_test = np.mean(abs(y_test - y_predict_test))

# YOUR CODE HERE
print("MAE on training data=", performance_train)
print("MAE on testing data=", performance_test)

# print(y.max())
# print(y.min())

MAE on training data= 44.14193729841633
MAE on testing data= 40.40741892504263


## Question 3
- Which performance metric did you use and why?
- Given the minimum and the maximum of `y` are 25 and 346, respectively, is this a good model for the dataset? Why or why not?
- Suggest some ways (at least one) that might improve the performance metric

YOUR ANSWER HERE

We used the Mean Absolute Error (MAE) performance metric because the MSE metric is not suitable in this case. 

No this is not a good model in respect to the maximum and minimum value of y. Because the errors are causing the prediction to derive a lot from the actual results. 

We might improve this performance by feature engineering. We might add interaction terms/polynomial terms/spines for better performance. 

# Classification

In [204]:
# Import the sonar dataset (contained in the sonar_data.csv file)
df = pd.read_csv("sonar_data.csv")

# Split it into X and y
X = df.drop(columns=['y']).values
y = df['y'].values
print("Shape of X -", X.shape)
print("Shape of y -", y.shape)
print("Unique values of y -", np.unique(y))

# Printing the first 5 rows
df.head()


Shape of X - (208, 60)
Shape of y - (208,)
Unique values of y - [0 1]


x1      x2       x3       x4      x5       x6      x7      x8       x9  \
0 -4.7000  2.9275  25.0272  10.4968 -3.1874  21.4930 -3.3071  4.5616  30.9744   
1 -4.3205  3.3075  26.0232  11.6536 -2.7523  22.2915 -2.6284  7.5696  31.3392   
2 -4.6070  3.4550  26.6376  12.5992 -3.1494  22.1400 -2.3259  8.0336  34.9568   
3 -4.8500  2.4275  25.4952  10.4920 -4.6105  21.1840 -3.7922  4.0416  26.9568   
4 -3.8570  3.6650  25.1544  10.9456 -3.8790  21.3245 -3.6701  5.9472  31.7024   

      x10  ...     x52     x53      x54      x55      x56      x57      x58  \
0  1.8999  ...  9.0405  5.1430  29.2703  20.1728  13.1336  16.1080  23.1764   
1  2.5848  ...  9.1260  5.1958  29.0816  20.2256  13.1528  16.0840  23.1029   
2  5.5746  ...  9.3480  5.3652  29.1615  20.4320  13.1952  16.1896  23.3444   
3  1.1376  ...  9.1815  5.0792  29.2550  20.2040  13.0584  16.0300  23.0924   
4  4.0131  ...  9.0465  5.1188  29.1785  20.2640  13.0120  16.0432  23.1008   

      x59     x60  y  
0  8.0630  3.0800  0  
1  8.0364  3.1100  0  
2  8.0665  3.1950  0  
3  8.0280  3.2925  0  
4  8.0749  3.2350  0  

[5 rows x 61 columns]

## Question 4
Why do we need normalization/standardization in this case?

YOUR ANSWER HERE

Here we need normalization because the feature values of X differ by a lot. If we predict without normalizing it, it will provide inaccurate results. 

In [205]:
# Use min-max scaling to normalize the features of X. Save the normalized array as X_norm. 
X_norm = preprocessing.minmax_scale(X)

# To confirm that it is now normalized, use the follwing code
df_norm = pd.DataFrame(X_norm, columns=df.columns[:-1])
df_norm.head()

# YOUR CODE HERE

x1        x2        x3        x4        x5        x6        x7  \
0  0.136431  0.156451  0.135677  0.035426  0.224956  0.237571  0.407468   
1  0.323009  0.221603  0.272011  0.150024  0.283033  0.666756  0.574405   
2  0.182153  0.246892  0.356110  0.243699  0.230028  0.585327  0.648810   
3  0.062684  0.070724  0.199737  0.034950  0.034999  0.071486  0.288149   
4  0.550885  0.282898  0.153088  0.079886  0.132640  0.147003  0.318182   

         x8        x9       x10  ...       x51       x52       x53       x54  \
0  0.340904  0.449282  0.285714  ...  0.231076  0.027104  0.155844  0.435673   
1  0.755458  0.483045  0.394537  ...  0.124502  0.108417  0.218182  0.111111   
2  0.819405  0.817859  0.869584  ...  0.032869  0.319544  0.418182  0.248538   
3  0.269239  0.077447  0.164593  ...  0.240040  0.161198  0.080519  0.409357   
4  0.531863  0.516659  0.621479  ...  0.155378  0.032810  0.127273  0.277778   

        x55       x56       x57       x58       x59       x60  
0  0.149660  0.417949  0.502841  0.185355  0.245179  0.060046  
1  0.199546  0.479487  0.389205  0.105263  0.140496  0.087760  
2  0.394558  0.615385  0.889205  0.368421  0.258953  0.166282  
3  0.179138  0.176923  0.133523  0.093822  0.107438  0.256351  
4  0.235828  0.028205  0.196023  0.102975  0.292011  0.203233  

[5 rows x 60 columns]

In [206]:
# Split the dataset (X_norm, y) into 85% training - 15% testing
# Then split the (X_train, y_train) again into 85% training and 15% validation.
# In both cases, use random_state=100
# Finally, print the shape of the X_train, X_valid, and X_test

X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=0.15, random_state=100)
X_train, X_valid, y_train, y_valid = train_test_split(X_train, y_train, test_size=0.15, random_state=100)

# YOUR CODE HERE
print("X_train shape = ", X_train.shape)
print("X_valid shape = ", X_valid.shape)
print("X_test shape = ", X_test.shape)

X_train shape =  (149, 60)
X_valid shape =  (27, 60)
X_test shape =  (32, 60)


## Quesion 5
We want to fit a Logistic Regression model. However, instead of manually setting the value of C (the inverse of the regularization strength), we want find the optimum value.
- On which data (train/test/valid) should we calculate the performance metric to find the optimum value of C?

YOUR ANSWER HERE

On "valid" data we should calculate the performance metric to find the optimum value of C.

In [207]:
# Complete the follwing code to find the optimum value of C
C_candidate = [0.01, 0.1, 1, 10]
performance_metric_all = {}

for C in C_candidate:
    # define the Logistic Regression model with C = C and random_state=100
    # Fit the model using X_traina and y_train
    # predict the model's output for the ____ dataset (which one should it be?)
    # calculate the performance metric for the ____ dataset
    # save it in the array performance_metric_all by using performance_metric_all[C] = per...
    model = linear_model.LogisticRegression(C=C, random_state=100)
    model = model.fit(X_train, y_train)
    y_predict_valid = model.predict(X_valid)
    performance_valid = np.mean(y_valid == y_predict_valid)*100
    performance_metric_all[C] = performance_valid
    

print(performance_metric_all)

# YOUR CODE HERE

{0.01: 62.96296296296296, 0.1: 70.37037037037037, 1: 66.66666666666666, 10: 66.66666666666666}


## Quesion 6
Based on the output of the previous block, which value of C should we choose?

YOUR ANSWER HERE

We should choose the "0.1" as the value of C as it gives the highest accuracy.

In [208]:
# Fit a logistic regression model on X_train, y_train using the C value found above. Use random_state=100
# predict the model's output for X_train and save it to y_predict_train
# predict the model's output for X_test and save it to y_predict_test

# YOUR CODE HERE
model = linear_model.LogisticRegression(C=0.1, random_state=100).fit(X_train, y_train)
y_predict_train = model.predict(X_train)
y_predict_test = model.predict(X_test)

In [209]:
# Calculate a suitable performance metric for the training set for this classification problem
# Calculate a suitable performance metric for the test set for this classification problem
# Print the values

from sklearn.metrics import f1_score

performance_train = f1_score(y_train, y_predict_train, average='binary')
performance_test = f1_score(y_test, y_predict_test, average='binary')

# YOUR CODE HERE
print("Training performance =", performance_train)
print("Testing performance =", performance_test)

Training performance = 0.8
Testing performance = 0.8947368421052632


## Question 7
- Which performance metric did you use and why?
- Is this a good model for the dataset? Why or why not?
- Suggest some ways (at least one) that might improve the performance metric

YOUR ANSWER HERE

I used the F1 measure for measuring performance. I have seen examples where accuracy miserably fails to judge the performance of a classifier. As F1 is a combination of precision and recall metric, I chose this one.

This is a mediocre model for the dataset. Because we know, the more the value of F1 measure is close to 1, the better the model is. 

We might improve the performance metric by tweaking the optimizer of the learning algorithm.